blockchain/peer_test.go

```diff
@@ -0,0 +1,216 @@
+package blockchain
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/p2p"
+)
+
+// These vars are used to record and verify different events generated during tests
+var (
+	numErrFuncCalls int        // number of calls to the errFunc
+	lastErr         error      // last generated error
+	peerTestMtx     sync.Mutex // needed as modifications of these variables are done from timer handler goroutine also
+)
+
+func resetErrors() {
+	peerTestMtx.Lock()
+	defer peerTestMtx.Unlock()
+	numErrFuncCalls = 0
+	lastErr = nil
+}
+
+func errFunc(err error, peerID p2p.ID) {
+	peerTestMtx.Lock()
+	defer peerTestMtx.Unlock()
+	_ = peerID
+	lastErr = err
+	numErrFuncCalls++
+}
+
+// check if peer timer is running or not (a running timer can be successfully stopped)
+// Note: it does stop the timer!
+func checkByStoppingPeerTimer(t *testing.T, peer *bpPeer, running bool) {
+	assert.NotPanics(t, func() {
+		stopped := peer.timeout.Stop()
```

Comment:
please avoid accessing internal state in tests as much as possible. otherwise it will be almost impossible to change the struct without breaking tests.